### ISSUE TITLE
Type of "this" in field initializers & initializer blocks must be the least upper bound of all constructor result annotations

### ISSUE THREAD

**Issue by t-rasmud** [typetools/checker-framework#2265]
In the lock checker, `@GuardedBy <: @GuardedByUnknown`. 
Consider the following example which has an initialization block.

```
import org.checkerframework.checker.lock.qual.GuardedByUnknown;
import org.checkerframework.checker.lock.qual.GuardedBy;

public class InitializerLub{
    {
        @GuardedBy InitializerLub field = this;
    }
    @GuardedByUnknown InitializerLub() {}
}
```

We expect the type of `this` to be the least upper bound of all constructor types (`@GuardedByUnknown` in this example). So, the checker framework must report an error at line `@GuardedBy InitializerLub field = this;` because `this` should have type `@GuardedByUnknown` and assigning this to type `@GuardedBy` is illegal.
However, the command `javac -processor tainting InitializerLub.java` doesn't produce any warning.
This suggests that the checker framework incorrectly annotates the type of `this` with the default qualifier in hierarchy instead of the lub of all constructor types.
